### PR TITLE
Do not include unreleased changelog in released versions.

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** infrastructure.

This PR introduces a new `configure.ml` variable specifying if the version that we are building is a released version. If this is the case, then the Changes chapter of the reference manual won't include the “Unreleased changes” initial section. The test-suite will also fail if this variable is set to `true` but there are remaining unreleased changelog entries, the idea being to remind the RM that all such entries should have been included in the appropriate section of the Changes chapter and removed from `doc/changelog` before proceeding to tag.

After putting the tag, it is expected that this variable will be turned back to `false`, although if it is not done immediately, it is not a problem, but as soon as the RM tries to backport a PR with a changelog entry, it will make the test-suite fail.

I documented this in the `release-process.md` document, and also took the opportunity to advise using `git push --tags --dry-run`.

- [X] Corresponding documentation was added / updated.

### CODE REVIEW

File: test-suite/misc/changelog.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
+
+for d in ../../doc/changelog/*; do
+  if [ -d "$d" ]; then
+    if [ "$(ls $d/*.rst | wc -l)" != "1" ]; then
+      echo "Fatal: unreleased changelog entries remain in ${d#../../}/"
+      echo "Include them in doc/sphinx/changes.rst and remove them from doc/changelog/"
+      exit 1
```
[the remark this answers is not in the document]
Comment (Zimmi48) [reply test-suite/misc/changelog.sh]: Right, I intended to have a mechanism to run this test only when `is_a_released_version` is `true` and then I forgot. What is strange to me is that when I ran the test-suite, it passed, even if when I tried to run it manually, it did fail.

[the remark this answers is not in the document]
Comment (SkySkimmer) [reply test-suite/misc/changelog.sh]: misc tests are run with current directory test-suite/, not test-suite/misc/

**Review by jfehrle — COMMENTED** [review]
Suppose I add a new tactic and document it.  IIUC, this proposal would facilitate hiding the new item in the changelog, but the documentation for the tactic would still be visible anyway.  It seems to me you'd want a user-visible tag on the added doc such as `(version 8.11)`.  And if we have those in the doc then why not use those in the changelist, too?  These tags could be kept around for a while.

Separately, would it make any sense to convert the checklist items in `release-process.md` into a script?  It might reduce operator errors.  On the other hand, we only do a handful of releases each year.

**Comment by Zimmi48** [comment]
> Suppose I add a new tactic and document it. IIUC, this proposal would facilitate hiding the new item in the changelog, but the documentation for the tactic would still be visible anyway. It seems to me you'd want a user-visible tag on the added doc such as `(version 8.11)`. And if we have those in the doc then why not use those in the changelist, too? These tags could be kept around for a while.

Nope. I'm not hiding anything. The main idea is to take advantage of the backporting process to never misplace changelog entries as it has happened often in the past. Another important side effect is never having to worry about merge conflicts in `CHANGES.md` anymore. More about the basic ideas behind this at: https://about.gitlab.com/2018/07/03/solving-gitlabs-changelog-conflict-crisis/

> Separately, would it make any sense to convert the checklist items in `release-process.md` into a script? It might reduce operator errors. On the other hand, we only do a handful of releases each year.

Indeed, that's part of my plan.

**Comment by Zimmi48** [comment]
Confirmed (https://gitlab.com/Zimmi48/coq/pipelines/60390418) that with the new version, setting the variable `is_a_released_version` to `true` without removing the unreleased changelog entries does make the test-suite fail. https://gitlab.com/Zimmi48/coq/pipelines/60404593 should then confirm that when `is_a_released_version` is set to `true`, removing all the unreleased changelog entries does make the test-suite succeed.

**Comment by ejgallego** [comment]
Could the check just be querying git to see if the current commit has a proper tag as an ancestor?

I'd be cool if we could program the logic and avoid the flag.

**Comment by Zimmi48** [comment]
I tried to avoid relying on git on purpose in this PR. We can always refine later though. In particular, I'm waiting to see what will be the proposed model for setting the version numbers after the full move to Dune (note that this is one of the main thing that has been delaying any action from me towards more automation: I don't want to have to do it twice).

My main concern with relying on a tag is that if the logic is broken, we only realize it once the tag is put, whereas in this case, the RM would open a PR setting the `is_a_released_version` to `true` while bumping the version numbers, and CI would confirm that the change is OK. Of course, if we were relying on standard Dune support, I would be less afraid.

**Comment by ejgallego** [comment]
> I tried to avoid relying on git on purpose in this PR. We can always refine later though. In particular, I'm waiting to see what will be the proposed model for setting the version numbers after the full move to Dune (note that this is one of the main thing that has been delaying any action from me towards more automation: I don't want to have to do it twice).

Makes sense, we could however take the opportunity as to add a comment detailing the TODO for the removal of the manual flag. IMHO the process should become fully automatic in the future, as long as it is manual prone we will always do it wrong.

**Comment by Zimmi48** [comment]
I don't think that TODO in code are so useful. There is already a few TODO in the release doc, and there is the [Release process automation project](https://github.com/orgs/coq/projects/2). I think it is enough for now, isn't it?

**Comment by ejgallego** [comment]
> I think it is enough for now, isn't it?

It not a blocker of course; my concrete fear of introducing this variable is that I'd like to know if it can be derived as a program formula of some kind. Then, I know it can be effectively eliminated later. The current docs don't clarify that.

**Comment by Zimmi48** [comment]
This is because I don't know what the process will look like with Dune yet. But even if it is not fully eliminated, as long as it is set and unset automatically, it should be fine, shouldn't it?

**Comment by Zimmi48** [comment]
FTR this PR has been ready for 5 days, and could have been merged anytime since.
@vbgl I'm not really sure what you intend to do for 8.10+beta1 if this is one is not included.

**Review by vbgl — APPROVED** [review]

**Pull request merged**